server/src/main/java/org/opensearch/index/translog/RemoteFsTranslog.java

```diff
@@ -316,13 +322,13 @@ private boolean prepareAndUpload(Long primaryTerm, Long generation) throws IOExc
     }
 
     private boolean upload(long primaryTerm, long generation, long maxSeqNo) throws IOException {
-        // During primary relocation (primary-primary peer recovery), both the old and the new primary have engine
-        // created with the RemoteFsTranslog. Both primaries are equipped to upload the translogs. The primary mode check
-        // below ensures that the real primary only is uploading. Before the primary mode is set as true for the new
-        // primary, the engine is reset to InternalEngine which also initialises the RemoteFsTranslog which in turns
-        // downloads all the translogs from remote store and does a flush before the relocation finishes.
-        if (primaryModeSupplier.getAsBoolean() == false) {
-            logger.debug("skipped uploading translog for {} {}", primaryTerm, generation);
+        // During primary relocation, both the old and new primary have engine created with RemoteFsTranslog and having
+        // ReplicationTracker.primaryMode() as true. However, before we perform the `internal:index/shard/replication/segments_sync`
+        // action which re-downloads the segments and translog on the new primary. We are ensuring 2 things here -
+        // 1. Using primaryModeSupplier, we prevent the new primary to do pre-emptive syncs
+        // 2. Using syncPermits, we prevent syncs at the desired time during primary relocation.
+        if (primaryModeSupplier.getAsBoolean() == false || syncPermits.tryAcquire(1) == false) {
```

Comment:
nit: avoid hard coding use `SYNC_PERMITS`

Reply:
With these new permits we are strictly serialising the uploads, could it have been possible to have concurrent uploads in the present world?

Reply:
Concurrent uploads per shard is not possible today due to the write lock and read lock acquired during the sync. The lock sequencing ensures that there is only 1 upload any time.

Reply:
yes, it is possible for concurrent uploads case. However i am not able to find a better way to allow that in the current flow.

Reply:
@gbbafna How are you deducing that there are concurrent uploads cases? The `upload` method in RemoteFsTranslog is only called by another instance method `prepareAndUpload`. For control to enter inside the method, it needs the write lock which is later downgraded to read lock. Still, as the read lock is acquired, no one else can acquire the write lock so any other upload needs to wait until the current upload is over. 

Let me know if there is any case where we can upload concurrently.